File: backend/app/api/routes/categories.py
"""Categories d'un espai. Cada espai te el seu pla, sense res compartit."""

from __future__ import annotations

from decimal import Decimal

from fastapi import APIRouter, HTTPException, Query, status
from sqlalchemy import func, select, update

from app.deps import DbSession, EditableWorkspace, Workspace
from app.models import Category, LlmSuggestion, Merchant, RecurringSeries, Rule, Transaction
from app.schemas.common import Message
from app.schemas.transaction import CategoryCreate, CategoryOut, CategoryUpdate
from app.services.seed import SLUG_INTERNAL_TRANSFER, SLUG_UNCATEGORIZED, slugify

router = APIRouter(prefix="/categories", tags=["categories"])

PROTECTED_SLUGS = {SLUG_UNCATEGORIZED, SLUG_INTERNAL_TRANSFER}


def _to_out(
    category: Category,
    *,
    transaction_count: int = 0,
    total_amount: Decimal = Decimal(0),
) -> CategoryOut:
    return CategoryOut(
        id=category.id,
        parent_id=category.parent_id,
        slug=category.slug,
        name=category.name,
        full_name=category.full_name,
        kind=category.kind,
        color=category.color,
        icon=category.icon,
        is_system=category.is_system,
        is_subscription=category.is_subscription,
        transaction_count=transaction_count,
        total_amount=total_amount,
    )


def _get_in_workspace(db: DbSession, workspace, category_id: int) -> Category:
    category = db.get(Category, category_id)
    if category is None or category.ledger_id != workspace.id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Categoria no trobada")
    return category


def _ensure_two_level_depth(db: DbSession, workspace, parent_id: int | None) -> None:
    if parent_id is None:
        return
    parent = _get_in_workspace(db, workspace, parent_id)
    if parent.parent_id is not None:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Nomes s'admeten dos nivells")


def _ensure_not_reparenting_parent(db: DbSession, workspace, category: Category) -> None:
    child_count = db.scalar(
        select(func.count(Category.id)).where(
            Category.ledger_id == workspace.id, Category.parent_id == category.id
        )
    )
    if child_count:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "No es pot canviar el pare d'una categoria que te subcategories",
        )


def _category_stats(db: DbSession, workspace) -> dict[int, tuple[int, Decimal]]:
    rows = db.execute(
        select(
            Transaction.category_id,
            func.count(Transaction.id),
            func.coalesce(func.sum(Transaction.amount), 0),
        )
        .where(Transaction.ledger_id == workspace.id, Transaction.category_id.is_not(None))
        .group_by(Transaction.category_id)
    ).all()
    return {category_id: (count, Decimal(str(total))) for category_id, count, total in rows}


def _rollup_stats(
    categories: list[Category], stats: dict[int, tuple[int, Decimal]]
) -> dict[int, tuple[int, Decimal]]:
    rolled: dict[int, tuple[int, Decimal]] = {}
    for category in categories:
        count, total = stats.get(category.id, (0, Decimal(0)))
        rolled[category.id] = (count, total)
    for category in categories:
        if category.parent_id is not None:
            parent_count, parent_total = rolled.get(category.parent_id, (0, Decimal(0)))
            child_count, child_total = rolled[category.id]
            rolled[category.parent_id] = (
                parent_count + child_count,
                parent_total + child_total,
            )
    return rolled


def _reassign_category_references(db: DbSession, workspace, from_id: int, to_id: int) -> None:
    db.execute(
        update(Transaction)
        .where(Transaction.ledger_id == workspace.id, Transaction.category_id == from_id)
        .values(category_id=to_id)
    )
    db.execute(
        update(Merchant)
        .where(Merchant.ledger_id == workspace.id, Merchant.default_category_id == from_id)
        .values(default_category_id=to_id)
    )
    db.execute(
        update(Rule)
        .where(Rule.ledger_id == workspace.id, Rule.set_category_id == from_id)
        .values(set_category_id=to_id)
    )
    db.execute(
        update(RecurringSeries)
        .where(RecurringSeries.ledger_id == workspace.id, RecurringSeries.category_id == from_id)
        .values(category_id=to_id)
    )
    db.execute(
        update(LlmSuggestion)
        .where(LlmSuggestion.suggested_category_id == from_id)
        .values(suggested_category_id=to_id)
    )


@router.get("", response_model=list[CategoryOut])
def list_categories(
    db: DbSession,
    workspace: Workspace,
    with_stats: bool = Query(default=False),
):
    categories = db.scalars(
        select(Category)
        .where(Category.ledger_id == workspace.id)
        .order_by(Category.kind, Category.position, Category.name)
    ).all()
    if not with_stats:
        return [_to_out(category) for category in categories]

    stats = _rollup_stats(categories, _category_stats(db, workspace))
    return [
        _to_out(
            category,
            transaction_count=stats.get(category.id, (0, Decimal(0)))[0],
            total_amount=stats.get(category.id, (0, Decimal(0)))[1],
        )
        for category in categories
    ]


@router.post("", response_model=CategoryOut, status_code=status.HTTP_201_CREATED)
def create_category(payload: CategoryCreate, db: DbSession, workspace: EditableWorkspace):
    parent = None
    if payload.parent_id is not None:
        parent = _get_in_workspace(db, workspace, payload.parent_id)
        _ensure_two_level_depth(db, workspace, payload.parent_id)

    base_slug = f"{parent.slug}-{slugify(payload.name)}" if parent else slugify(payload.name)
    slug = base_slug
    suffix = 2
    while db.scalar(
        select(Category).where(Category.ledger_id == workspace.id, Category.slug == slug)
    ):
        slug = f"{base_slug}-{suffix}"
        suffix += 1

    category = Category(
        ledger_id=workspace.id,
        slug=slug,
        name=payload.name,
        kind=parent.kind if parent else payload.kind,
        parent_id=parent.id if parent else None,
        color=payload.color,
        icon=payload.icon,
        is_subscription=payload.is_subscription,
    )
    db.add(category)
    db.commit()
    db.refresh(category)
    return _to_out(category)


@router.patch("/{category_id}", response_model=CategoryOut)
def update_category(
    category_id: int, payload: CategoryUpdate, db: DbSession, workspace: EditableWorkspace
):
    category = _get_in_workspace(db, workspace, category_id)
    data = payload.model_dump(exclude_unset=True)
    if (parent_id := data.get("parent_id")) is not None:
        if parent_id == category_id:
            raise HTTPException(
                status.HTTP_400_BAD_REQUEST, "Una categoria no pot ser pare d'ella mateixa"
            )
        _ensure_two_level_depth(db, workspace, parent_id)
        _ensure_not_reparenting_parent(db, workspace, category)
    for field, value in data.items():
        setattr(category, field, value)
    db.commit()
    db.refresh(category)
    return _to_out(category)


@router.delete("/{category_id}", response_model=Message)
def delete_category(
    category_id: int,
    db: DbSession,
    workspace: EditableWorkspace,
    reassign_to: int | None = Query(default=None),
):
    """Esborra una categoria. Si te moviments, cal passar reassign_to."""
    category = _get_in_workspace(db, workspace, category_id)

    if category.slug in PROTECTED_SLUGS:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST, "Aquesta categoria del sistema no es pot esborrar"
        )

    child_count = db.scalar(
        select(func.count(Category.id)).where(
            Category.ledger_id == workspace.id, Category.parent_id == category_id
        )
    )
    if child_count:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "Primer cal esborrar o moure les subcategories",
        )

    used = db.scalar(
        select(func.count(Transaction.id)).where(Transaction.category_id == category_id)
    )
    if used and reassign_to is None:
        raise HTTPException(
            status.HTTP_409_CONFLICT, f"Hi ha {used} moviments en aquesta categoria"
        )

    if reassign_to is not None:
        if reassign_to == category_id:
            raise HTTPException(
                status.HTTP_400_BAD_REQUEST, "No es pot reassignar a la mateixa categoria"
            )
        _get_in_workspace(db, workspace, reassign_to)
        _reassign_category_references(db, workspace, category_id, reassign_to)

    db.delete(category)
    db.commit()
    return Message(message="Categoria esborrada")

File: src/services/categories.ts
/**
 * Categories: tree, statistics and deletion with reassignment.
 *
 * A translation of `backend/app/api/routes/categories.py`, with the logic
 * taken out of the route and put here.
 */

import { and, count, eq, isNotNull, sum } from "drizzle-orm";

import { db } from "../db/client.ts";
import {
  categories,
  llmSuggestions,
  merchants,
  rules,
  transactions,
  type Category,
  type CategoryKind,
} from "../db/schema/index.ts";
import { AppError, ConflictError, NotFoundError } from "../lib/http.ts";
import { money, toMoneyString, type MoneyString } from "../lib/money.ts";
import { PROTECTED_SLUGS, slugify } from "./slugs.ts";

/** A category with what is shown of it on screen. */
export interface CategoryView {
  id: number;
  parentId: number | null;
  slug: string;
  name: string;
  /** «Parent › Child», like the Python's `full_name`. */
  fullName: string;
  kind: CategoryKind;
  color: string;
  icon: string;
  isSystem: boolean;
  transactionCount: number;
  totalAmount: MoneyString;
  /** One of those that can never be deleted. */
  isProtected: boolean;
}

export interface NodeCategory extends CategoryView {
  filles: CategoryView[];
}

/** The workspace's categories, in the order they should be shown. */
export async function listCategories(ledgerId: number): Promise<Category[]> {
  return db
    .select()
    .from(categories)
    .where(eq(categories.ledgerId, ledgerId))
    .orderBy(categories.kind, categories.position, categories.name);
}

/**
 * Transaction count and sum per category, with the children **rolled up into
 * the parent**, as `_rollup_stats` did.
 */
async function rollupStats(ledgerId: number): Promise<Map<number, [number, MoneyString]>> {
  const rows = await db
    .select({
      categoryId: transactions.categoryId,
      n: count(transactions.id),
      total: sum(transactions.amount),
    })
    .from(transactions)
    .where(and(eq(transactions.ledgerId, ledgerId), isNotNull(transactions.categoryId)))
    .groupBy(transactions.categoryId);

  const propies = new Map<number, [number, MoneyString]>();
  for (const row of rows) {
    if (row.categoryId === null) continue;
    propies.set(row.categoryId, [row.n, row.total ?? "0.00"]);
  }
  return propies;
}

function acumula(
  all: Category[],
  propies: Map<number, [number, MoneyString]>,
): Map<number, [number, MoneyString]> {
  const acumulades = new Map<number, [number, MoneyString]>();
  for (const category of all) {
    acumulades.set(category.id, propies.get(category.id) ?? [0, "0.00"]);
  }
  for (const category of all) {
    if (category.parentId === null) continue;
    const parent = acumulades.get(category.parentId);
    const filla = acumulades.get(category.id);
    if (!parent || !filla) continue;
    acumulades.set(category.parentId, [
      parent[0] + filla[0],
      toMoneyString(money(parent[1]).plus(money(filla[1]))),
    ]);
  }
  return acumulades;
}

/** The whole tree, grouped by type, with statistics if they are asked for. */
export async function categoryTree(
  ledgerId: number,
  withStats = true,
): Promise<Record<CategoryKind, NodeCategory[]>> {
  const all = await listCategories(ledgerId);
  const stats = withStats
    ? acumula(all, await rollupStats(ledgerId))
    : new Map<number, [number, MoneyString]>();

  const perId = new Map(all.map((c) => [c.id, c]));
  const view = (c: Category): CategoryView => {
    const [n, total] = stats.get(c.id) ?? [0, "0.00"];
    const parent = c.parentId === null ? undefined : perId.get(c.parentId);
    return {
      id: c.id,
      parentId: c.parentId,
      slug: c.slug,
      name: c.name,
      fullName: parent ? `${parent.name} › ${c.name}` : c.name,
      kind: c.kind,
      color: c.color,
      icon: c.icon,
      isSystem: c.isSystem,
      transactionCount: n,
      totalAmount: total,
      isProtected: PROTECTED_SLUGS.includes(c.slug),
    };
  };

  const tree: Record<CategoryKind, NodeCategory[]> = {
    expense: [],
    income: [],
    transfer: [],
  };
  for (const category of all) {
    if (category.parentId !== null) continue;
    tree[category.kind].push({
      ...view(category),
      filles: all.filter((c) => c.parentId === category.id).map(view),
    });
  }
  return tree;
}

/** A category of this workspace, or 404. */
export async function categoryInWorkspace(id: number, ledgerId: number): Promise<Category> {
  const [category] = await db
    .select()
    .from(categories)
    .where(and(eq(categories.id, id), eq(categories.ledgerId, ledgerId)))
    .limit(1);
  if (!category) throw new NotFoundError("Aquesta categoria no existeix");
  return category;
}

/** A slug unique within the workspace, adding `-2`, `-3`... if needed. */
async function freeSlug(ledgerId: number, base: string): Promise<string> {
  let candidat = base;
  let sufix = 2;
  for (;;) {
    const [xoc] = await db
      .select({ id: categories.id })
      .from(categories)
      .where(and(eq(categories.ledgerId, ledgerId), eq(categories.slug, candidat)))
      .limit(1);
    if (!xoc) return candidat;
    candidat = `${base}-${sufix}`;
    sufix += 1;
  }
}

export interface CreateCategory {
  name: string;
  kind: CategoryKind;
  parentId: number | null;
  color: string;
  icon: string;
}

export async function createCategory(
  ledgerId: number,
  data: CreateCategory,
): Promise<Category> {
  let parent: Category | null = null;
  if (data.parentId !== null) {
    parent = await categoryInWorkspace(data.parentId, ledgerId);
    // Two levels only: a category with a parent cannot have children.
    if (parent.parentId !== null) {
      throw new AppError("Nomes s'admeten dos nivells de categories", 422);
    }
  }

  const base = parent ? `${parent.slug}-${slugify(data.name)}` : slugify(data.name);
  const slug = await freeSlug(ledgerId, base);

  const [creada] = await db
    .insert(categories)
    .values({
      ledgerId,
      slug,
      name: data.name,
      // A subcategory always inherits the parent's type.
      kind: parent ? parent.kind : data.kind,
      parentId: parent?.id ?? null,
      color: data.color,
      icon: data.icon,
      isSystem: false,
      position: 0,
    })
    .returning();

  if (!creada) throw new AppError("No s'ha pogut crear la categoria", 500);
  return creada;
}

export async function renameCategory(
  id: number,
  ledgerId: number,
  name: string,
): Promise<Category> {
  await categoryInWorkspace(id, ledgerId);
  const [actualitzada] = await db
    .update(categories)
    .set({ name })
    .where(eq(categories.id, id))
    .returning();
  if (!actualitzada) throw new NotFoundError("Aquesta categoria no existeix");
  return actualitzada;
}

/** How many transactions there are in a category. */
export async function transactionsOf(categoryId: number): Promise<number> {
  const [row] = await db
    .select({ n: count() })
    .from(transactions)
    .where(eq(transactions.categoryId, categoryId));
  return row?.n ?? 0;
}

async function tefilles(id: number, ledgerId: number): Promise<number> {
  const [row] = await db
    .select({ n: count() })
    .from(categories)
    .where(and(eq(categories.ledgerId, ledgerId), eq(categories.parentId, id)));
  return row?.n ?? 0;
}

/**
 * Deletes a category.
 *
 * If it has transactions and no destination is given, it is a **409**: the
 * interface uses it to ask whoever it is to pick a destination category.
 *
 * Mind the order: everything has to be reassigned **before** deleting,
 * because the foreign key of `rules.set_category_id` is CASCADE and deleting
 * first would take away the rules that pointed at it.
 */
export async function deleteCategory(
  id: number,
  ledgerId: number,
  reassignTo: number | null,
): Promise<void> {
  const category = await categoryInWorkspace(id, ledgerId);

  if (PROTECTED_SLUGS.includes(category.slug)) {
    throw new AppError("Aquesta categoria del sistema no es pot esborrar", 422);
  }

  if ((await tefilles(id, ledgerId)) > 0) {
    throw new AppError("Primer cal esborrar o moure les subcategories", 422);
  }

  const used = await transactionsOf(id);
  if (used > 0 && reassignTo === null) {
    throw new ConflictError(
      `Hi ha ${used} ${used === 1 ? "moviment" : "moviments"} en aquesta categoria`,
      "Tria a quina categoria han d'anar a parar.",
    );
  }

  await db.transaction(async (tx) => {
    if (reassignTo !== null) {
      if (reassignTo === id) {
        throw new AppError("No es pot reassignar a la mateixa categoria", 422);
      }
      // It must belong to this workspace: otherwise transactions could be moved to another one.
      const [desti] = await tx
        .select({ id: categories.id })
        .from(categories)
        .where(and(eq(categories.id, reassignTo), eq(categories.ledgerId, ledgerId)))
        .limit(1);
      if (!desti) throw new NotFoundError("La categoria de desti no existeix");

      await tx
        .update(transactions)
        .set({ categoryId: reassignTo })
        .where(and(eq(transactions.ledgerId, ledgerId), eq(transactions.categoryId, id)));
      await tx
        .update(merchants)
        .set({ defaultCategoryId: reassignTo })
        .where(and(eq(merchants.ledgerId, ledgerId), eq(merchants.defaultCategoryId, id)));
      await tx
        .update(rules)
        .set({ setCategoryId: reassignTo })
        .where(and(eq(rules.ledgerId, ledgerId), eq(rules.setCategoryId, id)));
      // The recurring series of this category are deleted in cascade
      // (`fk_recurring_series_category_id_categories`). If the destination one
      // has the same pattern, `detectRecurring` proposes them again.
      await tx
        .update(llmSuggestions)
        .set({ suggestedCategoryId: reassignTo })
        .where(eq(llmSuggestions.suggestedCategoryId, id));
    }

    await tx.delete(categories).where(eq(categories.id, id));
  });
}

/**
 * The categories as options for a `<select>` with `<optgroup>`.
 *
 * This is what replaces the 372-line `SelectorCategoria`: two levels are
 * exactly what an `<optgroup>` knows how to do.
 */
export interface CategoryGroup {
  tag: string;
  options: { value: number; text: string }[];
}

export async function categoryOptions(
  ledgerId: number,
  excloure: readonly number[] = [],
): Promise<CategoryGroup[]> {
  const all = await listCategories(ledgerId);
  const fora = new Set(excloure);
  const groups: CategoryGroup[] = [];

  for (const parent of all) {
    if (parent.parentId !== null || fora.has(parent.id)) continue;
    const filles = all.filter((c) => c.parentId === parent.id && !fora.has(c.id));
    const options = [
      // The parent can be chosen too: some transactions belong to no child.
      { value: parent.id, text: parent.name },
      ...filles.map((f) => ({ value: f.id, text: `  ${f.name}` })),
    ];
    groups.push({ tag: parent.name, options });
  }
  return groups;
}

/** How many categories the workspace has. To know whether to seed the plan. */
export async function countCategories(ledgerId: number): Promise<number> {
  const [row] = await db
    .select({ n: count() })
    .from(categories)
    .where(eq(categories.ledgerId, ledgerId));
  return row?.n ?? 0;
}
